### PR TITLE
[HttpFoundation] Deprecate passing null as $requestIp in IpUtils

### PR DESCRIPTION
**Pull request by W0rma**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | https://github.com/symfony/symfony/issues/43350#issuecomment-936951604
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: UPGRADE-5.4.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ HttpKernel
 HttpFoundation
 --------------
 
+ * Deprecate passing `null` as `$requestIp` to `IpUtils::__checkIp()`, `IpUtils::__checkIp4()` or `IpUtils::__checkIp6()`, pass an empty string instead.
```
Comment (derrabus) [inline UPGRADE-5.4.md]:
```suggestion
 * Deprecate passing `null` as `$requestIp` to `IpUtils::checkIp()`, `IpUtils::checkIp4()` or `IpUtils::checkIp6()`, pass an empty string instead.
```

File: src/Symfony/Component/HttpClient/NoPrivateNetworkHttpClient.php [hunk 2]
```diff
@@ -80,7 +80,7 @@ public function request(string $method, string $url, array $options = []): Respo
 
         $options['on_progress'] = function (int $dlNow, int $dlSize, array $info) use ($onProgress, $subnets, &$lastPrimaryIp): void {
             if ($info['primary_ip'] !== $lastPrimaryIp) {
-                if (IpUtils::checkIp($info['primary_ip'], $subnets ?? self::PRIVATE_SUBNETS)) {
+                if (IpUtils::checkIp($info['primary_ip'] ?? '', $subnets ?? self::PRIVATE_SUBNETS)) {
```
Comment (derrabus) [inline src/Symfony/Component/HttpClient/NoPrivateNetworkHttpClient.php]:
```suggestion
                if ($info['primary_ip'] && IpUtils::checkIp($info['primary_ip'], $subnets ?? self::PRIVATE_SUBNETS)) {
```

File: src/Symfony/Component/HttpKernel/HttpCache/SubRequestHandler.php [hunk 3]
```diff
@@ -30,7 +30,7 @@ public static function handle(HttpKernelInterface $kernel, Request $request, int
         $trustedHeaderSet = Request::getTrustedHeaderSet();
 
         // remove untrusted values
-        $remoteAddr = $request->server->get('REMOTE_ADDR');
+        $remoteAddr = $request->server->get('REMOTE_ADDR', '');
         if (!IpUtils::checkIp($remoteAddr, $trustedProxies)) {
```
Comment (derrabus) [inline src/Symfony/Component/HttpKernel/HttpCache/SubRequestHandler.php]:
```suggestion
        $remoteAddr = $request->server->get('REMOTE_ADDR');
        if (!$remoteAddr || !IpUtils::checkIp($remoteAddr, $trustedProxies)) {
```

File: src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php [hunk 4]
```diff
@@ -40,7 +40,6 @@ public function getIpv4Data()
             [false, '1.2.3.4', '256.256.256/0'], // invalid CIDR notation
             [false, 'an_invalid_ip', '192.168.1.0/24'],
             [false, '', '1.2.3.4/1'],
-            [false, null, '1.2.3.4/1'],
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php:43]: Please move the deprecated test cases into a new test that also checks if the deprecation has been triggered.

**Review by derrabus — CHANGES_REQUESTED** [review]
Thank you for working on this topic!

Let's avoid the `checkIp()` if the IP to check is `null`. We already know the result.

File: src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php (hunk 4 above)
Comment (W0rma) [reply src/Symfony/Component/HttpFoundation/Tests/IpUtilsTest.php:43]: done

**Review by derrabus — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @W0rma.

**Pull request merged**